File: blogeur_ia_laravel/src/Services/BlogService.php
<?php

namespace BlogueurIa\Services;

use BlogueurIa\Models\BlogArticle;
use BlogueurIa\Support\Slugify;
use Illuminate\Support\Facades\DB;
use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;

class BlogService
{
    public function generateUniqueArticleSlug(string $titre): string
    {
        $base = Slugify::make($titre, 80);
        $candidate = $base;
        $counter = 2;
        while (BlogArticle::where('slug', $candidate)->exists()) {
            $candidate = "{$base}-{$counter}";
            $counter++;
        }
        return $candidate;
    }

    public function creerArticle(array $input): BlogArticle
    {
        $slug = $this->generateUniqueArticleSlug($input['titre']);
        $publier = $input['publier_immediatement'] ?? false;

        return BlogArticle::create([
            'slug' => $slug,
            'titre' => $input['titre'],
            'chapo' => $input['chapo'],
            'contenu_html' => $input['contenu_html'],
            'categorie' => $input['categorie'],
            'tags' => $input['tags'] ?? [],
            'auteur' => $input['auteur'],
            'image_couverture_url' => $input['image_couverture_url'] ?? null,
            'statut' => $publier ? 'publie' : 'brouillon',
            'source' => $input['source'] ?? 'manuel',
            'created_by' => $input['created_by'],
            'date_publication' => $publier ? now() : null,
        ]);
    }

    public function publierArticle(string $slug): BlogArticle
    {
        $article = BlogArticle::where('slug', $slug)->first();
        if (!$article) {
            throw new NotFoundHttpException('Article introuvable');
        }
        $article->statut = 'publie';
        $article->date_publication = $article->date_publication ?? now();
        $article->save();
        return $article;
    }

    public function getArticlePublie(string $slug): BlogArticle
    {
        $article = BlogArticle::where('slug', $slug)->where('statut', 'publie')->first();
        if (!$article) {
            throw new NotFoundHttpException('Article introuvable');
        }
        return $article;
    }

    public function listArticlesPublies(?string $categorie = null, int $limit = 20, int $offset = 0)
    {
        $query = BlogArticle::where('statut', 'publie');
        if ($categorie) {
            $query->where('categorie', $categorie);
        }
        return $query->orderByDesc('date_publication')->skip($offset)->take($limit)->get();
    }

    public function listArticlesSimilaires(string $slugActuel, int $limit = 4)
    {
        $actuel = BlogArticle::where('slug', $slugActuel)->first();
        if (!$actuel) {
            return collect();
        }

        $resultats = BlogArticle::where('statut', 'publie')
            ->where('slug', '!=', $slugActuel)
            ->where('categorie', $actuel->categorie)
            ->orderByDesc('date_publication')
            ->take($limit)
            ->get();

        if ($resultats->count() < $limit && !empty($actuel->tags)) {
            $deja = $resultats->pluck('id')->push($actuel->id)->all();
            $candidats = BlogArticle::where('statut', 'publie')
                ->whereNotIn('id', $deja)
                ->orderByDesc('date_publication')
                ->take(30)
                ->get();

            $scores = $candidats
                ->map(fn ($a) => ['a' => $a, 'score' => count(array_intersect($a->tags ?? [], $actuel->tags))])
                ->filter(fn ($x) => $x['score'] > 0)
                ->sortByDesc('score');

            foreach ($scores as $entry) {
                if ($resultats->count() >= $limit) {
                    break;
                }
                $resultats->push($entry['a']);
            }
        }

        if ($resultats->count() < $limit) {
            $deja = $resultats->pluck('id')->push($actuel->id)->all();
            $fallback = BlogArticle::where('statut', 'publie')
                ->whereNotIn('id', $deja)
                ->orderByDesc('date_publication')
                ->take($limit - $resultats->count())
                ->get();
            $resultats = $resultats->concat($fallback);
        }

        return $resultats->take($limit)->values();
    }

    public function supprimerArticle(string $slug): void
    {
        $article = BlogArticle::where('slug', $slug)->first();
        if (!$article) {
            throw new NotFoundHttpException('Article introuvable');
        }
        $article->delete();
    }

    public function listArticlesAdmin(?string $statut = null, int $limit = 20, int $offset = 0)
    {
        $query = BlogArticle::query();
        if ($statut) {
            $query->where('statut', $statut);
        }
        return $query->orderByDesc('created_at')->skip($offset)->take($limit)->get();
    }
}

File: blogeur_ia_python/blogueur_ia/blog_service.py
from datetime import datetime
from sqlmodel import select
from fastapi import HTTPException

from blogueur_ia.db import get_session
from blogueur_ia.models import BlogArticle
from blogueur_ia.utils import slugify


async def generate_unique_article_slug(titre: str) -> str:
    base = slugify(titre, max_len=80)
    candidate = base
    counter = 2
    async with get_session() as session:
        while True:
            exists = (await session.execute(
                select(BlogArticle).where(BlogArticle.slug == candidate)
            )).scalar_one_or_none()
            if not exists:
                return candidate
            candidate = f"{base}-{counter}"
            counter += 1


async def creer_article(
    titre: str, chapo: str, contenu_html: str, categorie: str,
    tags: list[str], auteur: str, created_by: str,
    image_couverture_url: str | None = None,
    source: str = "manuel",
    publier_immediatement: bool = False,
) -> BlogArticle:
    slug = await generate_unique_article_slug(titre)
    article = BlogArticle(
        slug=slug, titre=titre, chapo=chapo, contenu_html=contenu_html,
        categorie=categorie, tags=tags, auteur=auteur,
        image_couverture_url=image_couverture_url,
        statut="publie" if publier_immediatement else "brouillon",
        source=source, created_by=created_by,
        date_publication=datetime.utcnow() if publier_immediatement else None,
    )
    async with get_session() as session:
        session.add(article)
        await session.commit()
        await session.refresh(article)
    return article


async def publier_article(slug: str) -> BlogArticle:
    async with get_session() as session:
        article = (await session.execute(
            select(BlogArticle).where(BlogArticle.slug == slug)
        )).scalar_one_or_none()
        if not article:
            raise HTTPException(404, "Article introuvable")
        article.statut = "publie"
        article.date_publication = article.date_publication or datetime.utcnow()
        article.updated_at = datetime.utcnow()
        session.add(article)
        await session.commit()
        await session.refresh(article)
    return article


async def get_article_publie(slug: str) -> BlogArticle:
    async with get_session() as session:
        article = (await session.execute(
            select(BlogArticle).where(BlogArticle.slug == slug, BlogArticle.statut == "publie")
        )).scalar_one_or_none()
        if not article:
            raise HTTPException(404, "Article introuvable")
        return article


async def list_articles_publies(categorie: str | None = None, limit: int = 20, offset: int = 0) -> list[BlogArticle]:
    async with get_session() as session:
        stmt = select(BlogArticle).where(BlogArticle.statut == "publie")
        if categorie:
            stmt = stmt.where(BlogArticle.categorie == categorie)
        stmt = stmt.order_by(BlogArticle.date_publication.desc()).limit(limit).offset(offset)
        return (await session.execute(stmt)).scalars().all()


async def list_articles_similaires(slug_actuel: str, limit: int = 4) -> list[BlogArticle]:
    async with get_session() as session:
        actuel = (await session.execute(
            select(BlogArticle).where(BlogArticle.slug == slug_actuel)
        )).scalar_one_or_none()
        if not actuel:
            return []

        stmt = (
            select(BlogArticle)
            .where(BlogArticle.statut == "publie", BlogArticle.slug != slug_actuel,
                   BlogArticle.categorie == actuel.categorie)
            .order_by(BlogArticle.date_publication.desc()).limit(limit)
        )
        resultats = (await session.execute(stmt)).scalars().all()

        if len(resultats) < limit and actuel.tags:
            deja = {a.id for a in resultats} | {actuel.id}
            candidats = (await session.execute(
                select(BlogArticle).where(BlogArticle.statut == "publie", BlogArticle.id.not_in(deja))
                .order_by(BlogArticle.date_publication.desc()).limit(30)
            )).scalars().all()
            for a in sorted(candidats, key=lambda a: len(set(a.tags or []) & set(actuel.tags)), reverse=True):
                if len(resultats) >= limit:
                    break
                if set(a.tags or []) & set(actuel.tags):
                    resultats.append(a)

        if len(resultats) < limit:
            deja = {a.id for a in resultats} | {actuel.id}
            resultats += (await session.execute(
                select(BlogArticle).where(BlogArticle.statut == "publie", BlogArticle.id.not_in(deja))
                .order_by(BlogArticle.date_publication.desc()).limit(limit - len(resultats))
            )).scalars().all()

        return resultats[:limit]


async def supprimer_article(slug: str) -> None:
    async with get_session() as session:
        article = (await session.execute(
            select(BlogArticle).where(BlogArticle.slug == slug)
        )).scalar_one_or_none()
        if not article:
            raise HTTPException(404, "Article introuvable")
        await session.delete(article)
        await session.commit()


async def list_articles_admin(statut: str | None = None, limit: int = 20, offset: int = 0) -> list[BlogArticle]:
    async with get_session() as session:
        stmt = select(BlogArticle)
        if statut:
            stmt = stmt.where(BlogArticle.statut == statut)
        stmt = stmt.order_by(BlogArticle.created_at.desc()).limit(limit).offset(offset)
        return (await session.execute(stmt)).scalars().all()